pkg/binlog/reader/file.go

```diff
@@ -0,0 +1,181 @@
+	"time"
+
+	"github.com/pingcap/errors"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+)
+
+// FileReader is a binlog event reader which reads binlog events from a file.
+type FileReader struct {
+	mu sync.RWMutex
+	wg sync.WaitGroup
+
+	stage      readerStage
+	readOffset sync2.AtomicUint32
+	sendOffset sync2.AtomicUint32
+
+	parser *replication.BinlogParser
+	ch     chan *replication.BinlogEvent
+	ech    chan error
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// FileReaderConfig is the configuration used by a FileReader.
+type FileReaderConfig struct {
+	EnableRawMode bool
+	Timezone      *time.Location
+	ChBufferSize  int // event channel's buffer size
+	EchBufferSize int // error channel's buffer size
+}
+
+// FileReaderStatus represents the status of a FileReader.
+type FileReaderStatus struct {
+	Stage      string `json:"stage"`
+	ReadOffset uint32 `json:"read-offset"` // read event's offset in the file
+	SendOffset uint32 `json:"send-offset"` // sent event's offset in the file
+}
+
+// String implements Stringer.String.
+func (s *FileReaderStatus) String() string {
+	data, err := json.Marshal(s)
+	if err != nil {
+		log.Errorf("[FileReaderStatus] marshal status to json error %v", err)
+	}
+	return string(data)
+}
+
+// NewFileReader creates a FileReader instance.
+func NewFileReader(cfg *FileReaderConfig) Reader {
+	parser := replication.NewBinlogParser()
+	parser.SetVerifyChecksum(true)
+	parser.SetUseDecimal(true)
+	parser.SetRawMode(cfg.EnableRawMode)
+	if cfg.Timezone != nil {
+		parser.SetTimestampStringLocation(cfg.Timezone)
+	}
+	return &FileReader{
+		parser: parser,
+		ch:     make(chan *replication.BinlogEvent, cfg.ChBufferSize),
+		ech:    make(chan error, cfg.EchBufferSize),
+	}
+}
+
+// StartSyncByPos implements Reader.StartSyncByPos.
+func (r *FileReader) StartSyncByPos(pos gmysql.Position) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stage != stageNew {
+		return errors.Errorf("stage %s, expect %s, already started", r.stage, stageNew)
+	}
+
+	r.ctx, r.cancel = context.WithCancel(context.Background())
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		err := r.parser.ParseFile(pos.Name, int64(pos.Pos), r.onEvent)
+		if err != nil {
+			log.Errorf("[file reader] parse binlog file with error %s", errors.ErrorStack(err))
+			select {
+			case r.ech <- err:
+			case <-r.ctx.Done():
+			}
+		}
+	}()
+
+	r.stage = stagePrepared
+	return nil
+}
+
+// StartSyncByGTID implements Reader.StartSyncByGTID.
+func (r *FileReader) StartSyncByGTID(gSet gtid.Set) error {
+	// NOTE: may be supported later.
+	return errors.NotSupportedf("read from file by GTID")
+}
+
+// Close implements Reader.Close.
+func (r *FileReader) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stage == stageClosed {
+		return errors.New("already closed")
+	}
+
+	r.parser.Stop()
```

Comment:
it's better to put it after L135, even I know it's ok after I review its implemention

Reply:
addressed